add manual machine runner upgrade doc (#7909)
circleci/circleci-docs: jekyll/_cci2/runner-manually-upgrading-launch-agent.adoc

Author: done!

---
contentTags:
  platform:
  - Cloud
---
= Upgrading the CircleCI Machine Runner on Cloud
:page-layout: classic-docs
:page-liquid:
:page-description: Find resources to manually upgrade an already installed CircleCI Machine Runner on linux, windows, or macOS.
:icons: font
:toc: macro:
:toc-tittle:

> This page descrives how to update the CircleCI **machine runner** on CircleCI cloud

Reviewer: ~~ describes?

[#enable-auto-update]
== Enabling auto update

The most straightforward way to upgrade an existing CircleCI machine runner is to enable auto update in the configuration file and restart the runner instance. Auto update can be enabled by removing the `disable_auto_update` option from the machine runner configuration file, or by setting `runner.disable_auto_update` to false, then restarting machine runner instance either manually or by the process management used to manage the CircleCI machine runner agent. More details on configuring machine runner can be found in the xref:runner-config-reference[runner configuration reference] documentation.

In some cases an environment variable may have been used to disable the auto update functionality. If the machine runner installation is not updated after enabling the auto update functionality in the configuration file, the environment variable `LAUNCH_AGENT_RUNNER_DISABLE_AUTO_UPDATE` should be set to have a value of `false`, or unset entirely, and the machine runner instance restarted again. 

> === Linux 

Author: They refer back to "the process management used to manage the CircleCI machine runner agent". Should I remove the "Restart the machine runner instance either manually" and instead have the same note that's in the "[#manually-update-by-downloading-a-new-version-of-the-machine-runner-binary]" section?

> === Linux 

Reviewer 2: Ah, ok. In that case, yes perhaps we could add the note. I don't think it has to be in banner format. So, my suggestion would be something like this:

```
. Restart the machine runner instance with the process management used to manage the CircleCI machine runner agent. If process management is not being used to manage the machine runner, the instance will need to be manually stopped using `kill -9` on macOS and Linux systems, or the Task Manager on Windows systems, and then manually restarted. 
```

And if I am following the flow correctly, might I also suggest moving the "In some cases, an environment variable may have been used to disable the auto update functionality..." paragraph at the end of this section, after the Windows command? It might also work well to make it a banner by adding the `NOTE:` syntax. 

Only if this makes sense in terms of flow. I like the idea of having all of the pertinent information as close together as possible, and then a troubleshooting/related tip afterwards - at least that's how I understand the env var info..!

```shell
sudo systemctl restart circleci.service
```

=== macOS
```shell
sudo launchctl stop '/Library/LaunchDaemons/com.circleci.runner.plist'
sudo launchctl start '/Library/LaunchDaemons/com.circleci.runner.plist'
```

=== Windows
```shell
Stop-ScheduledTask -TaskName "CircleCI Launch Agent"
Start ScheduledTask -TaskName "CircleCI Launch Agent"
```

[#manually-download-updated-binary]
== Manually download the updated machine runner binary

> 

Reviewer 2: I think it would be great if we could lay out this section in a similar format as here: https://github.com/circleci/circleci-docs/blob/master/jekyll/_cci2/runner-upgrading-on-server.adoc#upgrading-machine-runner-in-a-non-containerized-environment

e.g. 
```
== Manually update by downloading a new version of the machine runner binary

The process ... can be broken down into three steps... These steps are expanded upon in the sections below.

. Download a new version of the binary
. Replace the existing binary with the new version

=== 1. Download a new version of the binary
...
==== Download the binary on Linux or macOS
...
==== Download the binary on Windows
...

=== 2. Replace the existing binary with the new version
...
==== Replace the binary on Linux
...
==== Replace the binary on macOS
...
==== Replace the binary on Windows
```

For machine runners where auto update is unable to be used, the machine runner binary, also sometimes known as the launch-agent binary,  can be manually downloaded and used to replace the existing version.

[#manually-download-mac-or-linux-binary]
=== Manually downloading the Linux or macOS machine runner agent

> As with the xref:runner-installation-linux.adoc[Linux] and xref:runner-installation-mac.adoc[macOS] installation process, the environment variable `platform` will need to be set to an appropriate value for the target installation system. 

Reviewer 2: ```suggestion
As with the xref:runner-installation-linux#[Linux] and xref:runner-installation-mac#[macOS] installation process, first you need to set the environment variable `platform` to the correct value for the target installation system. 
```
^ Different phrasing to make the instruction sound more definitive

chmod +x $file
```

[#manually-download-windows-binary]
=== Manually downloading the windows machine runner

```shell
# Download the new version of the CircleCI Agent manually on Windows with PowerShell
$Env:platform = "windows/amd64"
mkdir $Env:platform
$Env:baseURL =  "https://circleci-binary-releases.s3.amazonaws.com/circleci-launch-agent"
$Env:agentVersion = $(curl "$Env:baseURL/release.txt")
$Env:agentVersion = "$Env:agentVersion".Trim()
curl "$Env:baseURL/$Env:agentVersion/checksums.txt" -o checksums.txt
$Env:fileName = $(Select-String $Env:platform checksums.txt | cut -d ' ' -f 2 | sed 's/^.//')
$Env:fileName = $Env:fileName.Trim()
curl "$Env:baseURL/$Env:agentVersion/$Env:fileName" -o $Env:fileName
$Env:fileHash = $(Get-FileHash $Env:fileName -Algorithm "SHA256" | select -expand "Hash")
$Env:expectedHash = "$(grep $Env:filename checksums.txt)".Split(" ")[0]
if ($Env:expectedHash -ne $Env:fileHash) { Write-Output "CircleCI machine runner binary download failed, please try again"}else{Write-Output "CircleCI machine runner binary successfully downloaded to $CWD/$Env:fileName"}
```

== Manually replacing the machine runner binary

> Once the new binary has been successfully downloaded, the machine runner instance will need to be stopped, have the currently installed binary overwritten with the newly downloaded binary, and then be started again.

Reviewer 2: ```suggestion
Once the new binary has been successfully downloaded, you will need to stop the current machine runner instance, overwrite the currently installed binary with the newly downloaded binary, and then restart the machine runner instance.
```

[#manually-replace-linux-binary]
=== Linux 
```shell
sudo systemctl stop circleci.service
# Replace the current binary with the newly downloaded binary, by default the machine runner binary is installed to `/opt/circleci/circleci-launch-agent`
sudo systemctl start circleci.service
```

[#manually-replace=macos-binary]
=== macOS
```shell
sudo launchctl stop '/Library/LaunchDaemons/com.circleci.runner.plist'
# Replace the current binary with the newly downloaded binary, by default the machine runner binary is installed to `/opt/circleci/circleci-launch-agent`
sudo launchctl start '/Library/LaunchDaemons/com.circleci.runner.plist'
```

[#manually-replace-windows-binary]
=== Windows
```shell
Stop-ScheduledTask -TaskName "CircleCI Launch Agent"
# Replace the current binary with the newly downloaded binary, this will require renaming or deleting the existing binary
Start-ScheduledTask -TaskName "CircleCI Launch Agent"
```

> If process management is not being used to manage the machine runner, the process will need to be manually stopped using `kill -9` on macOS and Linux systems, or the Task Manager on Windows systems, and then manually restarted.  

Reviewer 2: ```suggestion
NOTE: If process management is not used to manage the machine runner, the process will need to be manually stopped using `kill -9` on macOS and Linux systems, or the Task Manager on Windows systems, and then manually restarted.  
```
I might also suggest moving this note earlier in this section, perhaps after the first paragraph "Once the new binary has been download..."
